diff --git a/audio/include/system/audio.h b/audio/include/system/audio.h
index 2f33b2b..76bd185 100644
--- a/audio/include/system/audio.h
+++ b/audio/include/system/audio.h
@@ -262,6 +262,7 @@ typedef enum {
     AUDIO_FORMAT_E_AC3               = 0x0A000000UL,
     AUDIO_FORMAT_DTS                 = 0x0B000000UL,
     AUDIO_FORMAT_DTS_HD              = 0x0C000000UL,
+    AUDIO_FORMAT_IEC61937            = 0x0D000000UL,
     AUDIO_FORMAT_MAIN_MASK           = 0xFF000000UL,
     AUDIO_FORMAT_SUB_MASK            = 0x00FFFFFFUL,
 
